### PR TITLE
Single Node Multiple Iteration Optimization

### PR DESCRIPTION
**Pull request by agnusmaximus**
- Uses the cyclades partitioning algorithm to partition sources into non-conflicting batches for multi iter optimization
- Target source elbo args share underlying variational parameters, which are updated multiple times in the process_sources function (by multiple threads). This function uses maximize_f to optimize the variational parameters.
- Modify one_node_infer and one_node_infer_multi to also return array of objective values
- Test case to ensure that cyclades is partitioning updates to non-conflict batches
- Test case to ensure that when optimizing sources, multi_iter gets to a higher sum objective values than both multi_iter with fewer iterations, and one_node_infer
- Todo: Add more test cases with overlapping sources to make sure multi_iter gets to a higher objective value

### CODE REVIEW

File: src/ParallelRun.jl
Comment (jeff-regier) [inline src/ParallelRun.jl]: We've been trying to use the "bang" notation for functions that modify their arguments. Since `components_tree` is being mutated here, perhaps call this `union_find!`?

File: src/ParallelRun.jl
Comment (jeff-regier) [inline src/ParallelRun.jl]: I agree, let's put the cyclades stuff in its own file (but not it's own module) and include it at the top of `ParallelRun.jl` if possible. We've being going all lower case for included files that aren't modules, e..g `src/cyclades.jl`

File: src/ParallelRun.jl
Comment (jeff-regier) [inline src/ParallelRun.jl]: I agree, `multi_iter` is a bit vague. How about `one_node_joint_infer`, since Cyclades optimizes all the light sources jointly, in contrast to `one_node_infer`? Or `one_node_infer_cyclades`?

File: src/ParallelRun.jl
Comment (jeff-regier) [inline src/ParallelRun.jl]: A bit cleaner: `floor(Int64, n_sources / nprocthreads)`

File: test/test_infer_multi_iter.jl
Comment (jeff-regier) [inline test/test_infer_multi_iter.jl]: One line `for` statements are a bit more compact: `for v_a in a, v_b in b`

File: test/test_infer_multi_iter.jl
Comment (jeff-regier) [inline test/test_infer_multi_iter.jl]: Though the unit tests often just get run with 1 thread. Maybe call this `n_simulated_threads` rather than `n_threads`? And rename `thread_id`  to `simulated_thread_id`, so no one gets confused?

File: test/test_infer_multi_iter.jl
Comment (jeff-regier) [inline test/test_infer_multi_iter.jl]: We generally use `@test` rather than `@assert` in the unit tests.

File: test/test_infer_multi_iter.jl
Comment (jeff-regier) [inline test/test_infer_multi_iter.jl]: `map` is a built-in function in Julia, so it would be more clear to use a variable different name.

File: test/test_infer_multi_iter.jl
Comment (jeff-regier) [inline test/test_infer_multi_iter.jl]: This looks like a really thorough test.

File: test/test_infer_multi_iter.jl
Comment (jeff-regier) [inline test/test_infer_multi_iter.jl]: Should this be `n_iters=2`? Even if not, it seems a bit confusing, though maybe I'm just not thinking about it the right way.

File: src/ParallelRun.jl
Comment (jeff-regier) [inline src/ParallelRun.jl]: Are all these `reserve_thread[] && ...` things necessary? @kpamnany --- Does Dtree need a free thread or something or what's the idea here? Presumably we'd rather use all the threads.

File: src/ParallelRun.jl
Comment (jeff-regier) [inline src/ParallelRun.jl]: Would it work to make `neighbor_map` a `Vector{Vector{Int64}}` rather than a `Dict`, by giving the sources consecutive integers as their names? And then storing the mapping for new names (i.e. the consecutive integers) to original names in a `Vector{Int64}`? We're try to avoid using `Dicts` whenever possible because they allocate a lot of memory. On the other hand, if this isn't a bottleneck if might not be worth changing.

File: src/ParallelRun.jl
Comment (jeff-regier) [inline src/ParallelRun.jl]: Oh, I see. Could Cyclades use `neighbors` directly without copying it? It's nice to avoid storing the same information twice, to simplify the code.

Comment (jeff-regier) [reply src/ParallelRun.jl]: Also maybe easier to think about the multi-node case with fewer data structures.

File: src/ParallelRun.jl
Comment (jeff-regier) [reply src/ParallelRun.jl]: Another thing about `Dicts` is the we don't have multi-node versions of them. But for arrays, we can create "global" or "distributed" versions that can be accessed from any node of the supercomputer.

File: src/ParallelRun.jl
Comment (jeff-regier) [inline src/ParallelRun.jl]: Do you how much of a speedup you get here for using 2,4, and 8 threads, rather than 1 thread? One issue we've had is that Julia's GC is, for now, single-threaded, so the speedup often isn't as good as we'd like. The overhead is particularly bad for parts of the code that allocate and free a lot of memory.

File: src/ParallelRun.jl
Comment (jeff-regier) [inline src/ParallelRun.jl]: I wouldn't, should be very fast single-threaded. Also, I think in the `src_div_par` branch we've stopped using a `Dict` for the results.

**Review by jeff-regier — CHANGES_REQUESTED** [review]
@agnusmaximus -- This all looks great, not just that it works, but that the code is so readable. I've requested some changes inline, mostly very minor. Thanks for the contribution!

**Comment by agnusmaximus** [comment]
Great! I'll fix these up sometime today and update all this

File: test/test_infer_multi_iter.jl
Comment (agnusmaximus) [reply test/test_infer_multi_iter.jl]: oops yeah it should be 2

File: src/ParallelRun.jl
Comment (agnusmaximus) [reply src/ParallelRun.jl]: Done

File: src/ParallelRun.jl
Comment (agnusmaximus) [reply src/ParallelRun.jl]: Done

File: src/ParallelRun.jl
Comment (agnusmaximus) [reply src/ParallelRun.jl]: Done

File: src/ParallelRun.jl
Comment (agnusmaximus) [reply src/ParallelRun.jl]: Done, using one_node_joint_infer

File: src/ParallelRun.jl
Comment (agnusmaximus) [reply src/ParallelRun.jl]: I see, it seems like a vector of InferResult is used instead of a map. Should I wait until that branch is merged with master to change it?

File: test/test_infer_multi_iter.jl
Comment (agnusmaximus) [reply test/test_infer_multi_iter.jl]: Done

File: test/test_infer_multi_iter.jl
Comment (agnusmaximus) [reply test/test_infer_multi_iter.jl]: Done

File: test/test_infer_multi_iter.jl
Comment (agnusmaximus) [reply test/test_infer_multi_iter.jl]: That's cool! Did not know you could do that. Done

File: test/test_infer_multi_iter.jl
Comment (agnusmaximus) [reply test/test_infer_multi_iter.jl]: Done

File: test/test_infer_multi_iter.jl
Comment (agnusmaximus) [reply test/test_infer_multi_iter.jl]: Done

File: src/ParallelRun.jl
Comment (agnusmaximus) [reply src/ParallelRun.jl]: Hm, it seems like doing the partitioning is relatively very fast compared to allocation and optimization (0.000236101 seconds vs 209.690328562s for 104 sources). I'm thinking maybe we won't need to parallelize the connected components code for single node because of this. What do you think?

File: src/ParallelRun.jl
Comment (agnusmaximus) [reply src/ParallelRun.jl]: Regarding the multi-node case, new code has to be written to right?

Comment (agnusmaximus) [reply src/ParallelRun.jl]: I think the neighbors array isn't copied, but a reference to it is stored in cyclades_neighbor_map

**Comment by agnusmaximus** [comment]
Let's not merge this just yet, I get a seg fault with 8/12 threads for some reason (4 threads seems to work fine... I'll need to check this out later)

**Comment by agnusmaximus** [comment]
Wait hmmm.... Running with 8 threads on the login nodes succeeds... I think it might be the way I'm running the batch script on the regular nodes that might be causing the seg fault... I'll look into that

**Comment by agnusmaximus** [comment]
Hm yeah, not sure why but sometimes with > 4 threads I get a seg fault: http://pastebin.com/eNw3nTxE. Has anything like this been seen before? It seems to be related to the garbage collector.

**Comment by kpamnany** [comment]
Do you get that segfault on a single node Max? How repeatable is it?

**Comment by gostevehoward** [comment]
In the scaling runs two weeks ago we saw a lot of segfaults but they only happened in 64+ node runs as I understand so we were never able to track down the root cause. If you have access to our slack channel take a look at the history for some more context. If you can provide detailed instructions on how to reliably reproduce the segfault with one (or few) nodes that might be of great value in tracking down the problem.

**Comment by agnusmaximus** [comment]
The segfaults seem to happen pretty often when I run with > 4 threads on the joint_infer test case. If you check out the multi_iter branch, and set num threads to be something like 12 ("export JULIA_NUM_THREADS=12") then "julia ./test/runtests.jl joint_infer" it should produce the error after running for like 10-15 minutes I think... (They happen pretty often, I have not gotten 12 threads to finish executing without running into the error that I linked in pastebin)

**Comment by agnusmaximus** [comment]
Update: Hmmm 12 threads just finished for me... weird. So it's not super reliable...
Update 2: Dang 6 threads just finished for me (on the login node), let me try it out using the batch script

**Comment by kpamnany** [comment]
I _think_ this is a race condition somewhere in Julia's runtime. And clearly, the particular interleaving that causes a problem doesn't happen frequently; the more the threads, the higher the probability. Good to know there's a single-node case that can produce it though. Please snapshot/tag the commit where you know it happens (and don't delete this branch please).

@tkelman: would you see if you can reproduce this/have Jameson take a look?

**Comment by jeff-regier** [comment]
@agnusmaximus  That segfault is a nightmare, at least if it turns out the be the same one we were stuck on during the IPDPS submission. @tkelman and @vtjnash are really good at fixing these segfaults if we can given them simple test cases that reliably exhibit the problem. So far we haven't had much luck isolating it though. If you have a single-node test case that might be the closest we've come.

**Comment by agnusmaximus** [comment]
I've pushed bin/batch_joint_infer.sl which just runs the joint_infer method test case on 8, and 12 (piping output to infer_multi_%d_thread). Usually one of these fails. Not sure how helpful this batch script will be in locating the segfault

**Comment by agnusmaximus** [comment]
Also regarding the segfault, I think Kiran is right in that it's some sort of race condition when accessing shared memory from multiple threads. If I remember right, when I used separate data per thread in multiple iteration training via cyclades (so threads did not share data) running on 1,8,12 threads finished fine. 

Edit: Actually, I have no idea... haha

**Review by jeff-regier — APPROVED** [review]

**Comment by jeff-regier** [comment]
Thanks for the revisions Max.

**Comment by agnusmaximus** [comment]
For batch_joint_infer.jl to run right, you need to add "const dt_nodeid = 1" on line 26 of ParallelRun.jl. Sorry about this!

**Pull request merged**
